### PR TITLE
Websocket transport support for v2ray

### PR DESCRIPTION
**Pull request by xiaokangwang**
Websocket transport implements a HTTP(S) compliable, surveillance proof transport method with plausible deniability.

This transport was disscussed at
https://github.com/v2ray/v2ray-core/issues/224
https://trello.com/c/3uCCeBkC/8-add-websocket-transport

### CODE REVIEW

File: transport/internet/ws/config_json.go [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+import (
+	"encoding/json"
+)
+
+func (this *Config) UnmarshalJSON(data []byte) error {
+	type JsonConfig struct {
+		ConnectionReuse bool   `json:"connectionReuse"`
+		Path            string `json:"Path"`
+		Pto             string `json:"Pto"`
```
Comment (v2ray) [inline transport/internet/ws/config_json.go]: What does PTO mean? Path?

File: transport/internet/ws/dialer.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+func wsDial(src v2net.Address, dest v2net.Destination) (*wsconn, error) {
+	//internet.DialToDest(src, dest)
+	commonDial := func(network, addr string) (net.Conn, error) {
+		return internet.DialToDest(src, dest)
+	}
+
+	dialer := websocket.Dialer{NetDial: commonDial, ReadBufferSize: 65536, WriteBufferSize: 65536}
+
+	effpto := func(dst v2net.Destination) string {
```
Comment (v2ray) [inline transport/internet/ws/dialer.go]: Unit test this function please.

File: transport/internet/ws/hub.go [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+			}
+		}
+		return
+
+	})
+
+	errchan := make(chan error)
+
+	listenerfunc := func() error {
+		return http.ListenAndServe(address.String()+":"+strconv.Itoa(int(port.Value())), nil)
```
Comment (v2ray) [inline transport/internet/ws/hub.go]: Don't use ListenAndServe, you can't stop it.

File: transport/internet/ws/ws.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+      "Path":"ws", // the path our ws handler bind
+      "Pto": "wss/ws", // the transport protocol we are using ws or wss(listen ws with tls)
+    }
+
+It is worth noting that accepting a non-valid cert is not supported as a self-signed or invalid cert can be a sign of a website that is not correctly configured and lead to additional investigation.
+
+
+This transport was disscussed at
+https://github.com/v2ray/v2ray-core/issues/224
+https://trello.com/c/3uCCeBkC/8-add-websocket-transport
```
Comment (v2ray) [inline transport/internet/ws/ws.go]: Not necessary to expose Trello link to public. It is not visible anyway.

File: transport/internet/ws/wsconn.go [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+		func1:=func(){
+		func1()
+	  }
+		won't work, failed to compile for it can't find func1.
+
+		Should following workaround panic,
+		readNext could have been called before the actual defination was made,
+		This is very unlikely.
+	*/
+	readNext := func(b []byte) (n int, err error) { panic("Runtime unstable. Please report this bug to developer.") }
```
Comment (v2ray) [inline transport/internet/ws/wsconn.go]: Move this as a member function to solve the recursive issue.

File: transport/internet/ws/wsconn.go [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+
+	"github.com/gorilla/websocket"
+)
+
+type wsconn struct {
+	wsc         *websocket.Conn
+	readBuffer  *bufio.Reader
+	connClosing bool
+	reusable    bool
+	rlock       *sync.Mutex
```
Comment (v2ray) [inline transport/internet/ws/wsconn.go]: Use sync.RWMutex to combine read and write lock.

File: transport/internet/ws/wsconn.go [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+		pongRcv <- 0
+		return nil
+	})
+
+	go func() {
+		for !ws.connClosing {
+			ws.wlock.Lock()
+			ws.wsc.WriteMessage(websocket.PingMessage, nil)
+			ws.wlock.Unlock()
+			tick := time.NewTicker(time.Second * 3)
```
Comment (v2ray) [inline transport/internet/ws/wsconn.go]: Use time.After()

File: transport/internet/ws/wsconn.go [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+	pongRcv := make(chan int, 0)
+	ws.wsc.SetPongHandler(func(data string) error {
+		pongRcv <- 0
+		return nil
+	})
+
+	go func() {
+		for !ws.connClosing {
+			ws.wlock.Lock()
+			ws.wsc.WriteMessage(websocket.PingMessage, nil)
```
Comment (v2ray) [inline transport/internet/ws/wsconn.go]: I think you are writing Ping messages too often. Try to sleep a while between each ping messages.

File: transport/internet/ws/hub.go [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+		log.Error("WS:WSListener->listenws->ListenAndServe!" + err.Error())
+	}
+
+	return err
+
+}
+
+func (wsl *WSListener) converttovws(w http.ResponseWriter, r *http.Request) (*wsconn, error) {
+	var upgrader = websocket.Upgrader{
+		ReadBufferSize:  65536,
```
Comment (v2ray) [inline transport/internet/ws/hub.go]: What is the default value if these are not set?

File: transport/internet/ws/config.go [hunk 10]
```diff
@@ -0,0 +1,21 @@
+package ws
+
+type Config struct {
+	ConnectionReuse bool
+	Path            string
+	Pto             string
+	Cert            string
```
Comment (v2ray) [inline transport/internet/ws/config.go]: Is there a way to reuse the TLS settings from upper level?

**Comment by v2ray** [comment]
In general:

1. Don't you nested functions. It is not testable.
2. Don't 'panic'.
3. Add some unit tests.

File: transport/internet/ws/config_json.go (hunk 1 above)
Comment (xiaokangwang) [inline transport/internet/ws/config_json.go]: protocol .....

File: transport/internet/ws/hub.go (hunk 9 above)
Comment (xiaokangwang) [inline transport/internet/ws/hub.go]: I have increased buffer size, this can increase performance

Comment (xiaokangwang) [inline transport/internet/ws/hub.go]: default value: 4096

File: transport/internet/ws/config.go (hunk 10 above)
Comment (xiaokangwang) [inline transport/internet/ws/config.go]: I don't think so, it seem tls settings are out of my code's scope.

**Comment by xiaokangwang** [comment]
Writing tests now.

File: transport/internet/ws/hub.go [hunk 11]
```diff
@@ -69,12 +70,25 @@ func (wsl *WSListener) listenws(address v2net.Address, port v2net.Port) error {
 	errchan := make(chan error)
 
 	listenerfunc := func() error {
-		return http.ListenAndServe(address.String()+":"+strconv.Itoa(int(port.Value())), nil)
+		ol, err := net.Listen("tcp", address.String()+":"+strconv.Itoa(int(port.Value())))
```
Comment (v2ray) [inline transport/internet/ws/hub.go]: Use ListenTCP here. It removes some unnecessary error checking in NewStoppableListener.

File: transport/internet/ws/stopablehttplistener.go [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+				continue
+			}
+		}
+
+		return newConn, err
+	}
+}
+
+func (sl *StoppableListener) Stop() {
+	close(sl.stop)
```
Comment (v2ray) [inline transport/internet/ws/stopablehttplistener.go]: You don't need the channel. Just use Listener.Close() here, and Accept() will return error when the listener is closed.

File: transport/internet/ws/hub.go (hunk 11 above)
Comment (xiaokangwang) [inline transport/internet/ws/hub.go]: Resolved by https://github.com/v2ray/v2ray-core/pull/238/commits/32e8deb6e8799b4e2f6dc5a0065f4a763dfef937

File: transport/internet/ws/ws_test.go [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+func Test_Connect_ws(t *testing.T) {
+	assert := assert.On(t)
+	(&Config{Pto: "ws", Path: ""}).Apply()
+	conn, err := Dial(v2net.AnyIP, v2net.TCPDestination(v2net.DomainAddress("echo.websocket.org"), 80))
+	assert.Error(err).IsNil()
+	conn.Write([]byte("echo"))
+	s := make(chan int)
+	go func() {
+		buf := make([]byte, 4)
+		conn.Read(buf)
```
Comment (v2ray) [inline transport/internet/ws/ws_test.go]: Please verify what is read.

**Pull request merged**